src/darwin/Framework/CHIP/MTRDeviceController.mm

```diff
@@ -654,6 +656,68 @@ - (void)setPairingDelegate:(id<MTRDevicePairingDelegate>)delegate queue:(dispatc
     });
 }
 
+- (void)setNocChainIssuer:(id<MTRNOCChainIssuer>)nocChainIssuer
+{
+    VerifyOrReturn([self checkIsRunning]);
+
+    dispatch_sync(_chipWorkQueue, ^{
+        VerifyOrReturn([self checkIsRunning]);
+
+        self->_operationalCredentialsDelegate->SetNocChainIssuer(nocChainIssuer);
+    });
+}
+
+- (NSNumber *)onNOCChainGenerationComplete:(NSData *)operationalCertificate
+                   intermediateCertificate:(NSData *)intermediateCertificate
+                           rootCertificate:(NSData *)rootCertificate
+                                       ipk:(NSData *)ipk
+                              adminSubject:(NSNumber *)adminSubject
+                                     error:(NSError * __autoreleasing *)error
+{
+    VerifyOrReturnValue([self checkIsRunning:error], [NSNumber numberWithUnsignedInt:CHIP_ERROR_INCORRECT_STATE.AsInteger()]);
+    VerifyOrReturnValue(operationalCertificate != nil, [NSNumber numberWithUnsignedInt:CHIP_ERROR_INVALID_ARGUMENT.AsInteger()]);
+    VerifyOrReturnValue(intermediateCertificate != nil, [NSNumber numberWithUnsignedInt:CHIP_ERROR_INVALID_ARGUMENT.AsInteger()]);
+    VerifyOrReturnValue(rootCertificate != nil, [NSNumber numberWithUnsignedInt:CHIP_ERROR_INVALID_ARGUMENT.AsInteger()]);
+
+    // use ipk and adminSubject from CommissioningParameters if not passed in
+    chip::Optional<chip::Controller::CommissioningParameters> commissioningParameters
+        = _cppCommissioner->GetCommissioningParameters();
```

Comment:
This needs to happen on the matter work queue, no?  That stuff is not threadsafe.

Reply:
In the latest revision of the PR, I now dispatch calls to the whole onNocChainGenerationCompleted method on the chipWorkQueue inside the block passed in to the NOCChainIssuer. So I believe, the quoted lines above will now execute on the Matter work queue.